File: apps/party-time-backend/src/main/java/com/partytime/service/AuthService.java
package com.partytime.service;

import com.partytime.api.dto.AccountRegisterDTO;
import com.partytime.api.dto.changepassword.ChangePasswordDTO;
import com.partytime.api.dto.login.LoginRequestDTO;
import com.partytime.api.dto.login.LoginResponseDTO;
import com.partytime.api.error.ApiError;
import com.partytime.configuration.PartyTimeConfigurationProperties;
import com.partytime.configuration.security.TokenAuthentication;
import com.partytime.jpa.entity.Account;
import com.partytime.jpa.repository.AccountRepository;
import com.partytime.mail.MailService;
import com.partytime.mail.model.verifyaccount.VerifyAccountModel;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class AuthService {

    private final AccountService accountService;
    private final AccountRepository accountRepository;
    private final PasswordEncoder passwordEncoder;
    private final MailService mailService;
    private final JwtService jwtService;
    private final PartyTimeConfigurationProperties configurationProperties;

    /**
     * F010 - Konto Erstellen
     * F014 - Konto Verifizieren
     */
    @Transactional
    public Account registerAccount(AccountRegisterDTO accountRegisterDTO) {
        if (accountRepository.existsByEmail(accountRegisterDTO.getEmail())) {
            // Account already exists!
            throw ApiError.badRequest("Ein Account mit dieser Email existiert bereits!")
                .asException();
        }
        Account account = Account.builder()
            .name(accountRegisterDTO.getName())
            .email(accountRegisterDTO.getEmail())
            .emailVerified(false)
            .pwHash(passwordEncoder.encode(accountRegisterDTO.getPassword()))
            .emailVerificationCode(UUID.randomUUID().toString())
            .build();
        Account savedAccount = accountRepository.save(account);

        // F014 - Konto Verifizieren
        mailService.sendMail(savedAccount.getEmail(), "Verifiziere deinen Account!",
            MailService.TEMPLATE_VERIFY_ACCOUNT, VerifyAccountModel.builder()
                .name(account.getName())
                    .homepage(configurationProperties.getUrl())
                .verificationLink(configurationProperties.getUrl() + "/auth/verify/" + account.getEmailVerificationCode())
                .build());

        log.info("Account created! Verification Code: " + account.getEmailVerificationCode());

        return savedAccount;
    }

    /**
     * F014 - Konto Verifizieren
     */
    @Transactional
    public void verifyAccount(String emailVerificationCode) {
        Account account = accountRepository.findByEmailVerificationCode(emailVerificationCode)
            .orElseThrow(() -> ApiError.badRequest("Email Verifizierung fehlgeschlagen").asException());
        account.setEmailVerified(true);
        account.setEmailVerificationCode(null);
        accountRepository.save(account);
    }

    /**
     * Implements F011
     */
    public LoginResponseDTO loginUser(LoginRequestDTO dto) {
        Account account = accountService.getAccount(dto.getEmail());
        if (!account.isEmailVerified()) {
            throw ApiError.forbidden().asException();
        }

        boolean passwordMatch = passwordEncoder.matches(dto.getPassword(), account.getPwHash());
        if (!passwordMatch) {
            throw ApiError.unauthorized().asException();
        }

        String accessToken = jwtService.createAccessToken(account);
        return new LoginResponseDTO(accessToken);
    }

    /**
     * Implements F013
     */
    @Transactional
    public void changePassword(ChangePasswordDTO changePasswordDTO, TokenAuthentication authentication) {
        String email = authentication.getPrincipal().getUsername();
        Account account = accountService.getAccount(email);
        boolean oldPasswordMatches = passwordEncoder.matches(changePasswordDTO.getOldPassword(), account.getPwHash());
        if (!oldPasswordMatches) {
            throw ApiError.unauthorized().asException();
        }
        String newPasswordHash = passwordEncoder.encode(changePasswordDTO.getNewPassword());
        account.setPwHash(newPasswordHash);
        accountRepository.save(account);
    }

}

File: apps/party-time-backend-kotlin/src/main/kotlin/com/partytime/service/AuthService.kt
package com.partytime.service

import com.partytime.api.dto.login.LoginRequestDTO
import com.partytime.api.dto.login.LoginResponseDTO
import com.partytime.api.error.ApiError
import com.partytime.api.error.asException
import com.partytime.configuration.security.AuthenticationToken
import com.partytime.jpa.entity.Account
import com.partytime.mail.model.MailEvent
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

private val authServiceLogger = KotlinLogging.logger("AuthService")
/**
 *  A [Service] class for authentication related functionality.
 *
 *  @param accountService used for retrieving Accounts and storing changed Account information
 *  @param cryptService Used to encode passwords and check if passwords match hashes
 *  @param jwtService Used to create [AuthenticationToken] for successfully logged-in users
 *  @param applicationEventPublisher Used to send [MailEvent] for verification requests
 *  @constructor Constructs a new [AuthService]
 */
@Service
class AuthService(
    private val accountService: AccountService,
    private val cryptService: CryptService,
    private val jwtService: JwtService,
    private val applicationEventPublisher: ApplicationEventPublisher
) {
    /**
     * F014 - Konto Verifizieren
     *
     * Verifies an account by e-mail-verification code
     * @param emailVerificationCode E-mail-verification code provided by the link the user clicked on
     */
    @Transactional
    fun verifyAccount(emailVerificationCode: String) {
        val account = accountService.accountByEmailVerificationCode(emailVerificationCode)
            .orElseThrow {
                ApiError.badRequest("E-Mail Verifizierung fehlgeschlagen").asException()
            }
        account.emailVerified = true
        account.emailVerificationCode = null
        accountService.saveAccount(account)
        authServiceLogger.info { "Account ${account.email} verified." }
    }

    /**
     * Implements F011
     *
     * Performs a login for the user by transforming their [LoginRequestDTO] into a [LoginResponseDTO].
     * The login will only succeed if the data contained in [LoginResponseDTO] can be associated with a
     *
     * - registered
     * - e-mail-verified
     * - password matching
     *
     * saved [Account].
     * @param dto Information provided by the user that wants to log in
     * @return Authentication token
     */
    fun loginUser(dto: LoginRequestDTO): LoginResponseDTO {
        val account = accountService.getAccountByMail(dto.email)
        if (!account.emailVerified) {
            throw ApiError.forbidden().asException()
        }

        val passwordMatch = cryptService.passwordMatchesHash(dto.password, account.pwHash)
        if (!passwordMatch) {
            throw ApiError.unauthorized().asException()
        }

        val accessToken = jwtService.createAccessToken(account)
        authServiceLogger.info { "Account ${account.email} JWT created." }
        return LoginResponseDTO(accessToken)
    }
}
